php-frontend/src/main/java/org/sonar/php/filters/SuppressWarningFilter.java

```diff
@@ -0,0 +1,124 @@
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.php.filters;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.sonar.plugins.php.api.tree.CompilationUnitTree;
+import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.declaration.AttributeGroupTree;
+import org.sonar.plugins.php.api.tree.declaration.AttributeTree;
+import org.sonar.plugins.php.api.tree.declaration.CallArgumentTree;
+import org.sonar.plugins.php.api.tree.expression.LiteralTree;
+import org.sonar.plugins.php.api.tree.lexical.SyntaxToken;
+import org.sonar.plugins.php.api.tree.lexical.SyntaxTrivia;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+public class SuppressWarningFilter extends PHPVisitorCheck implements FilterPhpIssue {
+
+  private final Map<String, Map<Integer, Set<String>>> suppressedRulesPerLinePerFile = new HashMap<>();
+  private String fileUri;
+
+  private static final String ARGUMENT_FORMAT = "\"[a-zA-Z0-9:]++\"";
+  private static final String ARGUMENTS_FORMAT = ARGUMENT_FORMAT + "(?:\\s*+,\\s*+"+ARGUMENT_FORMAT+")*+";
+  private static final Pattern SUPPRESS_WARNING_COMMENT_PATTERN = Pattern.compile("@SuppressWarnings\\s*+\\(\\s*+(?<arguments>"+ARGUMENTS_FORMAT+")\\s*+\\)");
+
+  public void reset() {
+    suppressedRulesPerLinePerFile.clear();
+  }
+
+  public void scanCompilationUnit(String fileUri, CompilationUnitTree tree) {
+    this.fileUri = fileUri;
+    super.visitCompilationUnit(tree);
+  }
+
+  @Override
+  public boolean accept(String fileName, String ruleName, int line) {
+    return !suppressedRulesPerLinePerFile.getOrDefault(fileName, Collections.emptyMap())
+      .getOrDefault(line, Collections.emptySet()).contains(ruleName);
+  }
+
+  @Override
+  public void visitAttribute(AttributeTree tree) {
+    Set<String> rulesSuppressed = tree.arguments().stream()
+      .map(CallArgumentTree::value)
+      // consider only string literal
+      .filter(expr -> expr.is(Tree.Kind.REGULAR_STRING_LITERAL)).map(LiteralTree.class::cast)
```

Comment:
Make sure there is no cast error

Reply:
There cannot be any: we are ensuring the object type through the call of `expr.is(Tree.Kind.REGULAR_STRING_LITERAL)`.
If we check the definition of the enum value `REGULAR_STRING_LITERAL` in `Tree.java`, we have this: `REGULAR_STRING_LITERAL(LiteralTree.class)`
This means that this enum value is tightly coupled to the class `LiteralTree`, so you don't need to also check it through `instanceof`.

You will encounter this mechanism in some other parsers (python and IaC for example) and also in checks code.

Reply:
Perfect, thanks for the clarification. Same as above, just wanted to make sure that this cast is thought about twice ;) (as i had some exceptions because of a similar case)